lab_3_generate_by_ngrams/main.py

```diff
@@ -160,6 +260,40 @@ def _postprocess_decoded_text(self, decoded_corpus: tuple[str, ...]) -> str | No
 
         In case of corrupt input arguments, None is returned
         """
+        if not isinstance(decoded_corpus, tuple) or not decoded_corpus:
+            return None
+        
+        text_with_spaces = ""
+        for token in decoded_corpus:
+            if token == self._end_of_word_token:
+                text_with_spaces += " "
+            else:
+                text_with_spaces += token
+        
+        words = text_with_spaces.split()
+        
+        if not words:
+            return None
+        
+        result_text = " ".join(words)
+        
+        first_letter_index = -1
+        for i, char in enumerate(result_text):
+            if char.isalpha(): 
+                first_letter_index = i
+                break
+        
+        if first_letter_index != -1:
+            before_letter = result_text[:first_letter_index]  
+            first_letter = result_text[first_letter_index]  
+            after_letter = result_text[first_letter_index + 1:]  
+            
+            result_text = before_letter + first_letter.upper() + after_letter
+        
+        if not result_text.endswith('.'):
+            result_text += '.'
+        
+        return result_text
```

Comment:
It is too complicated. Why do you need all this logic? Try to make it without `for` loops at all.